docs/src/proposals/comprehensive-compute-fees.md

```diff
@@ -0,0 +1,97 @@
+credited back after the message is fully processed.
+
+The goal of the fees is to cover the computation cost of processing a
+transaction.  Each of the above fee categories could be represented as a compute
+unit cost that, when added together, encompasses the entire cost of processing
+the transaction.  By calculating the total cost of the transaction the runtime
+can make better decisions on what transactions to process and when, as well as
+throw out transactions that exceed a cap.
+
+The per-compute unit fee doesn't have to be linear; developers could be
+non-linearly incentivized to reduce compute costs by optimizing in any of the 4
+fee categories listed above.
+
+To give developers some control over fees and the compute cap, a new built-in
+instruction could be introduced that requests a specific transaction-wide
+compute budget cap.  The instruction can be used by a developer to reduce the
+cap and thus fees they expect to pay, or to request a higher than the default
+cap.  The runtime could in-turn, use these instructions to determine how to
+schedule these more expensive transactions.
+
+### Transaction compute caps
+
+The current compute caps are independently applied to individual instructions.
+This means the overall transaction cap varies depending on how many instructions
+are in the transaction.  Instead, a transaction-wide cap is probably more
+appropriate.  One challenge of the transaction-wide cap is that each instruction
+(program) could no longer expect to be given an equal amount of compute units
+since the number of compute units will be based on the number of units already
+consumed by earlier instructions in the message.  This will provide some
+additional tuning and composability challenges for developers.
+
+### Builtin program compute costs
+
+Builtin programs do not incur any compute cost at the moment and are therefore
+not accounted for even when a bpf program invokes a builtin program via cpi
+(except the direct overhead cost of the cpi call itself).  Builtin programs
+should incur a program compute cost like any other program.  The challenge is
+how to quantify that compute cost.  One approach would be to measure the cost of
+each builtin program's instructions.
+
+## Proposed steps to implement
+
+- Apply the compute budget cap across the entire transaction rather than
+  per-instruction
+- Convert the per-sig fee to a per-sig compute cost and incorporate it into the
+  transaction-wide cap.  Initially, keep charging the current per-sig fee.
+- Add compute costs for write locks and account data loading and incorporate
+  those into the transaction-wide cap
+- Remove the per-sig centric fee and start charging fees based on the
+  transaction compute budget
+- Rework how the bank decides which transactions to process and when based on
+  the compute budget expectations
+- Add a built-in instruction that requests the amount of compute budget up front
+
+  ## Things to ponder
```

Comment:
I think devs would need an option into the `simulateTransaction` RPC method to figure the cost of a transaction, since part of the cost is now based on what code paths get executed